src/box/lua/config/applier/credentials.lua

```diff
@@ -1,30 +1,326 @@
-        log.verbose('credentials.apply: grant %s to %s %s (if not exists)',
-            privilege, role_or_user, name)
-        for _, permission in ipairs(privilege.permissions or {}) do
-            local opts = {if_not_exists = true}
-            if privilege.universe then
-                grant_f(name, permission, 'universe', nil, opts)
-            end
-            -- TODO: It is not possible to grant a permission for
-            -- a non-existing object. It blocks ability to set it
-            -- from a config. Disabled for now.
-            --[[
-            for _, space in ipairs(privilege.spaces or {}) do
-                grant_f(name, permission, 'space', space, opts)
+--[[
+This intermediate representation is a formatted set of
+all permissions for a user or role. It is required to
+standardize diff functions. All the validation is done
+by config or box.info(), so neither the format nor the
+helper function don't perform it. Below you can find two
+converters to this representation, from box format and
+config schema format, accordingly `privileges_{box,config}_convert()`.
+
+[obj_type][obj_name] = {
+    read = true,
+    write = true,
+    ...
+}
+
+obj_types:
+ - 'user'
+ - 'role'
+ - 'space'
+ - 'function'
+ - 'sequence'
+ - 'universe'
+
+obj_names:
+ - mostly user defined strings, provided by config or box
+ - special value '', when there is no obj_name, e.g. for
+   'universe' obj_type or for granting permission for all
+   objects of a type.
+
+privs:
+ - read
+ - write
+ - execute
+   - lua_eval
+   - lua_call
+   - sql
+ - session
+ - usage
+ - create
+ - drop
+ - alter
+ - reference
+ - trigger
+ - insert
+ - update
+ - delete
+
+Examples:
+- - box.schema.user.grant('myuser', 'execute', 'function', 'myfunc')
+  - ['function']['myfunc']['execute'] = true
+  - grant execute of myfunc
+
+- - box.schema.user.grant('myuser', 'execute', 'function')
+  - ['function']['']['execute'] = true
+  - grant execute of all registered functions
+
+- - box.schema.user.grant('myuser', 'read', 'universe')
+  - ['universe']['']['read'] = true
+  - grant read to universe
+
+- - box.schema.user.grant('myuser', 'execute', 'role', 'super')
+  - ['role']['super']['execute'] = true
+  - equivalent to granting a role to myuser
+
+]]--
+
+local function privileges_box_convert(privileges)
+    privileges = privileges or {}
+    assert(type(privileges) == 'table')
+
+    local res = {
+        ['user'] = {},
+        ['role'] = {},
+        ['space'] = {},
+        ['function'] = {},
+        ['sequence'] = {},
+        ['universe'] = {},
+    }
+
+    for _, priv in ipairs(privileges) do
+        local perms, obj_type, obj_name = unpack(priv)
+        obj_name = obj_name or ''
+
+        res[obj_type][obj_name] = res[obj_type][obj_name] or {}
+
+        for _, perm in ipairs(perms:split(',')) do
+            res[obj_type][obj_name][perm] = true
+        end
+    end
+
+    return res
+end
+
+local function privileges_config_convert(config_data)
```

Comment:
Regarding names. The usual convention is either `<component>_<action>` or just `<action>`. The `<action>` is usually a verb or a verb + a noun group. There are also constructors like `<component>_from_<something>`. We can use this naming scheme.

```lua
-- Create a privileges definition represented in the intermediate
-- representation described at beginning of the source file from
-- `box.schema.{user,role}.info(name)` data.
local function priv_ir_from_box(data)
    <...>
end

-- Create a privileges definition represented in the intermediate
-- representation described at beginning of the source file from
-- the declarative config data.
local function priv_ir_from_config(data)
    <...>
end
```

Reply:
We've discussed naming verbally and came to the following agreement:
1) `privileges_` prefix for all helper functions working with intermediate representation (besides `sync_privileges()` because it is part of applier logic more than ir).
2) `privileges_box_convert()` -> `privileges_from_box`
3) `privileges_config_convert()` -> `privileges_from_config`
4) `privileges_lack(current, target)` -> `privileges_subtract(target, current)`